src/core/qgsogrutils.cpp

```diff
@@ -274,6 +275,55 @@ QgsGeometry QgsOgrUtils::ogrGeometryToQgsGeometry( OGRGeometryH geom )
   if ( !geom )
     return QgsGeometry();
 
+  QgsWkbTypes::Type wkbType = static_cast<QgsWkbTypes::Type>( OGR_G_GetGeometryType( geom ) );
+  switch ( QgsWkbTypes::flatType( wkbType ) )
+  {
+    case QgsWkbTypes::Point:
+    {
+      // optimised case for point -- avoid wkb conversion
+      double x, y, z, m;
+      OGR_G_GetPointZM( geom, 0, &x, &y, &z, &m );
+      std::unique_ptr< QgsPoint > p = qgis::make_unique< QgsPoint >( wkbType, x, y, z, m );
+      return QgsGeometry( std::move( p ) );
+    }
+
+    case QgsWkbTypes::LineString:
+    {
+      // optimised case for line -- avoid wkb conversion
+      int count = OGR_G_GetPointCount( geom );
+      QVector< double > x;
+      x.resize( count );
+      QVector< double > y;
+      y.resize( count );
+      QVector< double > z;
+      double *pz = nullptr;
+      if ( QgsWkbTypes::hasZ( wkbType ) )
+      {
+        z.resize( count );
+        pz = z.data();
+      }
+      double *pm = nullptr;
+      QVector< double > m;
+      if ( QgsWkbTypes::hasM( wkbType ) )
+      {
+        m.resize( count );
+        pm = m.data();
+      }
+      double *px = x.data();
+      double *py = y.data();
+      double zz, mm;
+      for ( int i = 0; i < count; ++i )
```

Comment:
Using OGR_G_GetPointsZM() to directly fill the arrays should probably be faster

Reply:
Thanks -- updated.

Reply:
@rouault Are the docs for OGR_G_GetPointsZM incorrect? They state:

    pabyX a buffer of at least (sizeof(double) * nXStride * nPointCount) bytes, may be NULL.

Shouldn't this be just:

    pabyX a buffer of at least (nXStride * nPointCount) bytes, may be NULL.

Reply:
@nyalldawson You're right. The docs were incorrect. I've just fixed then

Reply:
Thanks!